NEW Nested gridfield

forsdahl opened the pull request:
<!--
  Thanks for contributing, you're awesome! ⭐

  Please read https://docs.silverstripe.org/en/contributing/code/ if you haven't contributed to this project recently.
-->
## Description
<!--
  Please describe expected and observed behaviour, and what you're fixing.
  For visual fixes, please include tested browsers and screenshots.
-->
Provides a `GridFieldComponent` for displaying a nested `GridField`, e.g. for `has_many` relations of the main record.

## Manual testing steps
<!--
  Include any manual testing steps here which a reviewer can perform to validate your pull request works correctly.
  Note that this DOES NOT replace unit or end-to-end tests.
-->


Can be tested in a standard Silverstripe 5.1 installation in `SecurityAdmin` for editing groups, by adding the following extension to the project and adding it to `SecurityAdmin` via yaml config:

```php
<?php

namespace Creamarketing\NestedGridField;

use SilverStripe\Core\Extension;
use SilverStripe\Security\Group;
use Symbiote\GridFieldExtensions\GridFieldNestedForm;

class SecurityAdminExtension extends Extension
{
    public function updateGridFieldConfig($config)
    {
        if ($this->owner->getModelClass() === Group::class) {
            // shows nested groups
            $config->addComponent(GridFieldNestedForm::create());
            // Shows members in groups
            // $config->addComponent(GridFieldNestedForm::create()->setRelationName('Members'));
        }
    }
}
```

```yml
SilverStripe\Admin\SecurityAdmin:
  extensions:
    - Creamarketing\NestedGridField\SecurityAdminExtension
```

## Issues
<!--
  List all issues here that this pull request fixes/resolves.
  If there is no issue already, create a new one! You must link your pull request to at least one issue.
-->
- #383

## Pull request checklist
<!--
  PLEASE check each of these to ensure you have done everything you need to do!
  If there's something in this list you need help with, please ask so that we can assist you.
-->
- [x] The target branch is correct
    - See [picking the right version](https://docs.silverstripe.org/en/contributing/code/#picking-the-right-version)
- [x] All commits are relevant to the purpose of the PR (e.g. no debug statements, unrelated refactoring, or arbitrary linting)
    - Small amounts of additional linting are usually okay, but if it makes it hard to concentrate on the relevant changes, ask for the unrelated changes to be reverted, and submitted as a separate PR.
- [x] The commit messages follow our [commit message guidelines](https://docs.silverstripe.org/en/contributing/code/#commit-messages)
- [x] The PR follows our [contribution guidelines](https://docs.silverstripe.org/en/contributing/code/)
- [ ] Code changes follow our [coding conventions](https://docs.silverstripe.org/en/contributing/coding_conventions/)
- [ ] This change is covered with tests (or tests aren't necessary for this change)
- [ ] Any relevant User Help/Developer documentation is updated; for impactful changes, information is added to the changelog for the intended release
- [x] CI is green

forsdahl committed:
NEW Initial nested gridfield component

forsdahl committed:
Styling for nested GridFields without title.
Added loading indicator while nested GridField is being fetched.

forsdahl committed:
Added initial unit test for GridFieldNestedForm

forsdahl committed:
Added initial user documentation for Nested GridFields.

GuySartorelli reviewed (changes requested):
I haven't done a comprehensive review of this, but here are some changes or questions I have so far:

- I think breadcrumbs in the GridField should include the parent record
    - In this case `Groups > Content Authors > Smith`
    ![image](https://github.com/symbiote/silverstripe-gridfieldextensions/assets/36352093/628b3901-20e1-4431-9f0f-851798fcbf34)
- Search/filter functionality in nested gridfield doesn't seem to work
- Nested GridField state is not retained - e.g. if I apply a sort to the nested gridfield, click to edit a record, and then click back - the sort order is back to its default state.
- Seems to be no way to edit the nested GridField except through extensions
    - The acceptance criterion "the configuration for the nested GridField can be customised by a developer if needed" it a bit vague but IMO it is important to be able to customise it for *each* `GridField` independently, not just as a global extension
- There's an acceptance criterion in the issue about an upper limit for recursion, but I can't find where that's implemented - has that been implemented?
- "When a record doesn't have any nested relations, nested gridfield can be configured not to show."
    - Has this been implemented? If so please update the documentation to mention this.
    - Similarly if there are other options that haven't been documented, please document them so that I can test them and other developers can make use of them.
- The approach of subclassing `GridFieldDetailForm` seems a bit unusual, it's definitely not what I was expecting. Can you please briefly explain the reasons behind that decision?
- Please update the component list in the README.


I have been testing this by adding a nested gridfield to groups in `SecurityAdmin` with this code:

```php
<?php

namespace App\NestedExtensions;

use SilverStripe\Core\Extension;
use SilverStripe\Forms\GridField\GridFieldConfig;
use Symbiote\GridFieldExtensions\GridFieldNestedForm;

class NestedGroups extends Extension
{
    public function updateGridFieldConfig(GridFieldConfig $config)
    {
        $config->addComponent(GridFieldNestedForm::create()->setRelationName('Members'));
    }
}
```

```yml
SilverStripe\Admin\SecurityAdmin:
  extensions:
    - App\NestedExtensions\NestedGroups
```

forsdahl committed:
Renamed GridFieldNestedForm_ItemRequest, to conform with coding styles.
Also added some php-docs for some public functions.

forsdahl committed:
Changed base-class of GridFieldNestedForm, it doesn't share much with
GridFieldDetailForm anymore.
Also changed how Breadcrumbs work.

forsdahl commented:
I now changed how Breadcrumbs work in nested GridFields. It now shows all the parent records, but the back link goes to the top-level instead of the parent record, since having to click back through all parent records to get back to the place you where is a bit confusing.

The reason `GridFieldNestedForm` extended `GridFieldDetailForm` are some legacy stuff, which isn't really true anymore in Silverstripe 5, so I now changed the base class to `AbstractGridFieldComponent` instead. `GridFieldNestedFormItemRequest` still extends `GridFieldDetailForm_ItemRequest` though, since those two actually share a lot of functionality.

The reason the search component doesn't work is probably due to the fact that the name of the nested GridFields contain "[" and "]" characters. The reason for this is so that `GridFieldEditableColumns` will work in a non-hacky way for nested GridFields. I'll see if I can come up with a solution that works for both components, or if I can find a fix for the search component working event with "["-characters in the GridField names.

I will implement some more UnitTests still, and look at the missing things for the acceptance criterias.

forsdahl committed:
Fixed linting issues

forsdahl committed:
Changed naming schema for nested GridFields, to not include [ or ]
characters. This makes them work also with the search component.

forsdahl committed:
Added one more unit test for GridFieldNestedForm

forsdahl committed:
Added configurable max nesting level for nested GridFields

ChloeSartorelli commented on src/GridFieldNestedForm.php:
```diff
@@ -0,0 +1,435 @@
+    {
+        $level = 0;
+        $c = $gridField->getForm()->getController();
+        while ($c && $c instanceof GridFieldDetailForm_ItemRequest) {
```
```suggestion
        while ($c && $c instanceof GridFieldNestedFormItemRequest) {
```
We only need to count nested gridfields - it doesn't matter if there are 2000 regular gridfield edit forms leading up to the root nested gridfield, those shouldn't be counted.

ChloeSartorelli commented on src/GridFieldNestedForm.php:
```diff
@@ -0,0 +1,435 @@
+            $level++;
+        }
+        return $level;
+    }
```
It is theoretically possible for an infinite loop of controllers referencing each other here - since we only need the nesting level to check if we're hitting the max, this method should be adjusted to return a boolean, which is whether we've hit the max nesting level or not. Something like this:

```php
private function atMaxNestingLevel(GridField $gridfield): bool
{
    $level = 0;
    $controller = $gridField->getForm()->getController();
    $maxLevel = $this->getMaxNestingLevel();
    while ($level < $maxLevel  && $controller && $controller instanceof GridFieldNestedFormItemRequest) {
        $controller = $controller->getController();
        $level++;
    }
    return $level >= $maxLevel;
}
```

ChloeSartorelli commented on src/GridFieldNestedForm.php:
```diff
@@ -0,0 +1,435 @@
+
+    public function getMaxNestingLevel()
+    {
+        return $this->maxNestingLevel ?: $this->config()->max_nesting_level;
```
```suggestion
        return $this->maxNestingLevel ?: static::config()->get('max_nesting_level');
```
Config is a static method so should be accessed statically.
The get method should be explicitly used instead of relying on the magic `__get()`, to make it clearer what's happening.

If there's any other usage of config in this PR, please also update those.

ChloeSartorelli commented on src/GridFieldNestedForm.php:
```diff
@@ -0,0 +1,435 @@
+        'handleNestedItem'
+    ];
+
+    private static $max_nesting_level = 10;
```
```suggestion
    private static $default_max_nesting_level = 10;
```
This value is used as a default, so we should be explicit about that.

Please also add a PHPDoc for this since it will display in the API documentation. Same with any config you've added in this PR that isn't inherited.

ChloeSartorelli commented on src/GridFieldNestedForm.php:
```diff
@@ -0,0 +1,435 @@
+    protected $relationName = 'Children';
+    protected $inlineEditable = false;
+    protected $canExpandCheck = null;
+    protected $maxNestingLevel = null;
```
Please add strict typehints for these (including the config properties), and for all methods that aren't being inherited from superclasses or interfaces. For all new API, we always add strict typehints as best practice.
Please see https://docs.silverstripe.org/en/5/contributing/coding_conventions/php_coding_conventions/

ChloeSartorelli replied:
For all of these which have a `getter` and `setter` method, please make the property `private`. There's no reason for subclasses to be accessing the properties directly.

ChloeSartorelli commented on src/GridFieldNestedForm.php:
```diff
@@ -0,0 +1,435 @@
+     */
+    public function getGridField()
+    {
+        return $this->gridField;
```
```suggestion
    public function getGridField(): GridField
    {
        return $this->gridField;
```
Please add strict typehints for arguments and return types for all methods in this PR which aren't inherited from superclasses or interfaces.

Note that PHPDocs are only required for parameters and return types if there is additional information, such as a description. If it's just the type, we don't need the PHPDoc when there's a real typehint.

ChloeSartorelli commented on src/GridFieldNestedForm.php:
```diff
@@ -0,0 +1,435 @@
+        }
+        return $level;
+    }
+
+    public function getColumnMetadata($gridField, $columnName)
+    {
+        return ['title' => ''];
+    }
+
+    public function getColumnsHandled($gridField)
+    {
+        return ['ToggleNested'];
+    }
+
+    public function getColumnAttributes($gridField, $record, $columnName)
+    {
+        return ['class' => 'col-listChildrenLink grid-field__col-compact'];
+    }
+
+    public function augmentColumns($gridField, &$columns)
+    {
+        if (!in_array('ToggleNested', $columns)) {
+            array_splice($columns, 0, 0, 'ToggleNested');
+        }
+    }
+
+    public function getColumnContent($gridField, $record, $columnName)
+    {
+        $nestingLevel = $this->getNestingLevel($gridField);
+        if ($nestingLevel >= $this->getMaxNestingLevel()) {
+            return '';
+        }
+        $gridField->addExtraClass('has-nested');
+        if ($record->ID && $record->exists()) {
+            $this->gridField = $gridField;
+            $this->record = $record;
+            $relationName = $this->getRelationName();
+            if (!$record->hasMethod($relationName)) {
+                return '';
+            }
```
This strikes me as a time we should be throwing an exception. If the developer has said "this relation exists", we shouldn't fail silently if we discover it doesn't actually exist.

ChloeSartorelli replied:
You marked this as resolved but didn't make any change, nor did you reply. Is there a reason to not throw an exception here? If not please make this change.

ChloeSartorelli commented on src/GridFieldNestedForm.php:
```diff
@@ -0,0 +1,435 @@
+
+    /**
+     * Set a callback to check which items in this grid that should show the expand link
+     * for nested gridfields. The callback should return a boolean value.
```
The usage of `canExpandCheck` makes it clear that the name of a method on the parent record can also be passed in. Please update the PHPDoc to make it clear this is possible.

ChloeSartorelli commented on src/GridFieldNestedForm.php:
```diff
@@ -0,0 +1,435 @@
+     * for nested gridfields. The callback should return a boolean value.
+     * @param callable $checkFunction
+     */
+    public function setCanExpandCheck($checkFunction)
```
```suggestion
     */
    public function setCanExpandCheck(callable|string $checkFunction): static
```

ChloeSartorelli commented on src/GridFieldNestedForm.php:
```diff
@@ -0,0 +1,435 @@
+     */
+    public function setCanExpandCheck($checkFunction)
+    {
+        $this->canExpandCheck = $checkFunction;
```
```suggestion
    public function setCanExpandCallback($callback)
    {
        $this->canExpandCallback = $callback;
```
Just making it clearer by the naming what this actually is - the property holds a callback (or the name of a method).

ChloeSartorelli commented on src/GridFieldNestedForm.php:
```diff
@@ -0,0 +1,435 @@
+                    return '';
+                } elseif (is_string($this->canExpandCheck)
+                    && $record->hasMethod($this->canExpandCheck)
+                    && !$this->record->{$this->canExpandCheck}($record)
```
- We don't need the `is_string()` check since the setter method can only accept a string or a callable.
- We should throw an exception if the record doesn't have that method, since that scenario means the developer has misconfigured this component. Not throwing an exception is just silently failing.

forsdahl replied:
Actually, the way the code works now it is needed. The first if statement will return false if it is a callable but the result from this callable is true (i.e. we can expand nested items on this record). So either it should be as it is, or it needs to be refactored a bit.

ChloeSartorelli replied:
Yup good point, I must not have had my coffee when I looked at this bit :p

ChloeSartorelli commented on src/GridFieldNestedForm.php:
```diff
@@ -0,0 +1,435 @@
+     */
+    public function getHTMLFragments($field)
+    {
+        if (DataObject::has_extension($field->getModelClass(), Hierarchy::class)) {
```
```suggestion
        if (is_a($field->getModelClass(), DataObject::class, true) && DataObject::has_extension($field->getModelClass(), Hierarchy::class)) {
```
Don't assume the modelclass is a `DataObject` subclass.
Please check the PR for other scenarios where that assumption has been made.

forsdahl replied:
Also added some checks to not assume the underlying list is Filterable, i.e. don't assume it has `filter` or `byID` functions.

ChloeSartorelli commented on src/GridFieldNestedForm.php:
```diff
@@ -0,0 +1,435 @@
+    /**
+     * @param GridField $field
+     */
+    public function getHTMLFragments($field)
```
The name of this method implies it should be returning something. Have you missed a return statement? Or are we (ab)using this method for something it wasn't intended for?

If the latter, at a minimum please include a code statement explaining what's going on, but preferably find a more suitable place to put the logic.

forsdahl replied:
This function is used because of the move between gridfields functionality. That functionality currently needs the move-to-parent url as a data-attribute on the grid field. This attribute needs to be added before anything is rendered for the gridfield, and it is also not dependent on individual records, so `getColumnContent` is also not optimal. The optimal place to do it seems to be `getURLHandlers`, but that is called too late, i.e. the added attribute will not be rendered. Added a comment about this in the code.

ChloeSartorelli commented on src/GridFieldNestedForm.php:
```diff
@@ -0,0 +1,435 @@
+        }
+    }
+
+    public function handleMoveToParent(GridField $gridField, $request)
```
Please add a PHPDoc comment to this and to all other public methods in this PR which aren't inherited from superclasses or interfaces.

ChloeSartorelli commented on src/GridFieldNestedForm.php:
```diff
@@ -0,0 +1,435 @@
+                                $orderableRows->handleReorder($gridField, $request);
+                            } catch (Exception $e) {
+                            }
+                        }
```
This level of nested control statements is a bit of a code smell - it's quite hard to follow the flow of logic at a glance. Please try to restructure this so that we're exiting early where possible. It may be appropriate to refactor some of this logic into private methods instead of having so many nested control statements.

forsdahl replied:
You are completely right, this code is not good. The logic is a bit complicated here, since there are quite a few possible cases for the moving functionality, but I now at least refactored it to use early return instead of this deep control statements structure.

ChloeSartorelli commented on src/GridFieldNestedFormItemRequest.php:
```diff
@@ -0,0 +1,173 @@
+use Symbiote\GridFieldExtensions\GridFieldEditableColumns;
+use Symbiote\GridFieldExtensions\GridFieldOrderableRows;
+
+class GridFieldNestedFormItemRequest extends GridFieldDetailForm_ItemRequest
```
Please add a PHPDoc comment to this and to all public methods in this class which aren't inherited from the superclass.

ChloeSartorelli commented on src/GridFieldNestedFormItemRequest.php:
```diff
@@ -0,0 +1,173 @@
+
+class GridFieldNestedFormItemRequest extends GridFieldDetailForm_ItemRequest
+{
+    
```
```suggestion
```

ChloeSartorelli commented on src/GridFieldNestedFormItemRequest.php:
```diff
@@ -0,0 +1,173 @@
+        return Controller::join_links($this->component->Link($this->record->ID), $action);
+    }
+    
+    public function ItemEditForm()
```
Please refactor this a bit so that the nested gridfield's `GridFieldConfig` is passed in from the `GridFieldNestedForm` component.
Then, please add a getter and setter method on `GridFieldNestedForm` so that developers have direct access to the config which will be used in the nested gridfield.

i.e. it should be possible to do this (feel free to change the method names):
```php
$nestedGridfieldComponent = new GridFieldNestedForm();
$parentGridFieldConfig->addComponent($nestedGridfieldComponent);

// alter the nested gridfield config
$nestedConfig = $nestedGridfieldComponent->getNestedConfig();
$nestedConfig->// add or remove components to suit business needs

// completely replace the nested gridfield config
$nestedGridfieldComponent->setNestedConfig($myNestedConfig);
```

forsdahl replied:
One reason the gridfield config is created where it is now is that in the default config, we sometimes want to check permissions on the specific record to see what components should be added. For example the default config is based on `GridFieldConfig_RecordEditor`, but if the user can't edit a given record, then that record's nested gridfield will not include the add button. This won't be possible in the same way if we move the logic to `GridFieldNestedForm`, since then the config is the same for all records. For now we have found it more usable to us to configure the nested gridfield config directly in the DataObject classes, or via extensions to said classes.
I can refactor it this way, but then we need to deside if we do any of these permission checks at all, since they might then be confusing for the developer.

ChloeSartorelli replied:
Ahh I've just noticed the `getNestedConfig()` check there which solves some of this. That should definitely be added to the documentation (which still needs to be added to as per https://github.com/symbiote/silverstripe-gridfieldextensions/pull/384#pullrequestreview-2013562592 "Please add additional documentation so I know what needs to be manually tested and so that developers know what functionality this thing has.").

That's probably okay for now, given there is at least a per-model level of control.

ChloeSartorelli commented on src/GridFieldNestedFormItemRequest.php:
```diff
@@ -0,0 +1,173 @@
+        $gridField->addExtraClass('nested');
+        $form = new Form($this, 'ItemEditForm', $fields, new FieldList());
+        $form->setStrictFormMethodCheck(false);
+        $form->disableSecurityToken();
```
Why are we disabling the security token and strict method check?

forsdahl replied:
This is some legacy stuff that doesn't seem to be needed anymore (probably dates back to before Silverstripe 4). I now removed these, all functionality seem to work without these, and better to leave them on than disabling them.

ChloeSartorelli commented on templates/Symbiote/GridFieldExtensions/GridFieldNestedForm.ss:
```diff
@@ -0,0 +1,6 @@
+<a class="btn btn-secondary btn--no-text btn--icon-large <% if $Toggle == 'open' %>font-icon-down-dir<% else %>font-icon-right-dir<% end_if %> cms-panel-link list-children-link" data-pjax-target="$PjaxFragment" href="$Link" data-toggle="$ToggleLink"></a>
```
Do I understand correctly that this "link" is used to collapse/expand the nested gridfield? If so please use the appropriate `aria` attributes for accessibility.
https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-expanded should set you on the right track.

ChloeSartorelli replied:
Also please use a `button` element instead of a link if possible. Buttons are more semantically correct here (and don't have the unexpected behaviour when middle clicking)

forsdahl replied:
Good points, much better with a button with aria-expanded instead of an anchor link.

GuySartorelli reviewed (changes requested):
I haven't done a full thorough code review but I have taken a look and highlighted some areas that don't meet our coding standards or which seemed like they could be improved.

I've also done another in-browser test and these are the problems I encountered:

- There's some unexpected state conflict between the nested gridfield and the main gridfield. Steps to reproduce:
    1. Have at least two items in the main gridfield, both with nested gridfields
    1. expand both nested gridfields
    1. Filter the main gridfield (e.g. search for "author") so that there's only one item in the main gridfield.
    1. Filter the nested gridfield.
    1. Click on a record in the nested gridfield, then click the back arrow in the breadcrumbs
    1. Remove the filter on the *main* gridfield
    1. Click on a record in the nested gridfield, then click the back arrow in the breadcrumbs
    1. Notice that the filter (which you cleared just before) is back again.
- Calling `setMaxNestingLevel(3)` on the nested gridfield component doesn't seem to work - though changing the configuration does.
    - My guess is we need to pass the nesting level down recursively to all child nested gridfield components. We should also disallow developers setting the value in nested components directly, so we don't get a parent with a max nesting level of 3 and a child with a max nesting level of 10 (as a random example).
    - I would have expected the nested gridfield settings to be inherited automatically, since this was tested with a `Hierarchy` object (using the example in the PR description). It looks like the reason this didn't happen is because the nested gridfield in this case is set up in `GridFieldNestedFormItemRequest` as a fresh instance of `GridFieldNestedForm()` which doesn't look at the settings defined in `getCMSFields()` at all.
- When trying to use sort to move items between nested gridfields, I moved a child group to be a sibling to what was its parent group. This resulted in a popup that said "An error occured while fetching data from the server Please try again later." - and an error in the network tab of the developer console which said `ERROR [UNKNOWN TYPE, ERRNO 400]`. The full error with stack trace was:
    ```text
    ERROR [UNKNOWN TYPE, ERRNO 400]: 
    IN POST /admin/security/groups/EditForm/field/groups/nested/1/ItemEditForm/field/groups-GridFieldNestedForm-1/nested/3/ItemEditForm/field/groups-GridFieldNestedForm-1-GridFieldNestedForm-3/nested/5/ItemEditForm/field/groups-GridFieldNestedForm-1-GridFieldNestedForm-3-GridFieldNestedForm-5/reorder?gridState-groups-0=%7B%22GridFieldOrderableRows%22%3A%7B%22enabled%22%3Atrue%7D%2C%22GridFieldNestedForm%22%3A%7B%22SilverStripe-Security-Group-2-Children%22%3Anull%2C%22SilverStripe-Security-Group-1-Children%22%3A1%7D%7D&gridState-groups-GridFieldNestedForm-1-1=%7B%22GridFieldOrderableRows%22%3A%7B%22enabled%22%3Atrue%7D%2C%22GridFieldNestedForm%22%3A%7B%22SilverStripe-Security-Group-3-Children%22%3A1%7D%7D&gridState-groups-GridFieldNestedForm-1-GridFieldNestedForm-3-2=%7B%22GridFieldOrderableRows%22%3A%7B%22enabled%22%3Atrue%7D%2C%22GridFieldNestedForm%22%3A%7B%22SilverStripe-Security-Group-4-Children%22%3Anull%2C%22SilverStripe-Security-Group-5-Children%22%3A1%7D%7D
    Line  in 
    
    Trace
    =====
    SilverStripe\Logging\DetailedErrorFormatter->output(400, , , , )
    DetailedErrorFormatter.php:55
    
    SilverStripe\Logging\DetailedErrorFormatter->format(Array)
    HTTPResponse.php:417
    
    SilverStripe\Control\HTTPResponse->outputBody()
    HTTPResponse.php:346
    
    SilverStripe\Control\HTTPResponse->output()
    index.php:25
    ```
    - It's worth noting that the group *did* move to where it was meant to move to, despite the error.
    
Additionally, please do these actions I previously requested in https://github.com/symbiote/silverstripe-gridfieldextensions/pull/384#pullrequestreview-2007670453
- Please add additional documentation so I know what needs to be manually tested and so that developers know what functionality this thing has.
- Please update the component list in the readme

forsdahl committed:
Fixed moving nested gridfield items to other gridfields

forsdahl committed:
Refactored nested GridField move to parent functionality.

forsdahl committed:
Fixes and some refactoring for max nesting level handling in
nested gridfields.

forsdahl committed:
PHPDoc additions and linting fixes for gridfield nested form

forsdahl committed:
Don't assume records are DataObjects in nested gridfields.
Also don't assume the list is filterable.

forsdahl committed:
Removed legacy disabling of security token and strict form method check
for nested gridfields, doesn't seem to be needed anymore.

forsdahl committed:
Added phpdoc for nested grid field item request handler class.

forsdahl committed:
Throw exception in nested gridfields if the relation is invalid.

forsdahl committed:
Changed some PHPDoc return types to real typehings in nested gridfield.

forsdahl committed:
Refactored grid field nested form link to be a button with aria-attributes,
for better accessibility.

ChloeSartorelli commented on src/GridFieldNestedForm.php:
```diff
@@ -0,0 +1,517 @@
+    
+    /**
+     * @var string
+     */
+    private $name;
+
+    /**
+     * @var bool
+     */
+    private $expandNested = false;
+
+    /**
+     * @var bool
+     */
+    private $forceCloseNested = false;
+
+    /**
+     * @var GridField
+     */
+    private $gridField = null;
+
+    /**
+     * @var string
+     */
+    private $relationName = 'Children';
+
+    /**
+     * @var bool
+     */
+    private $inlineEditable = false;
+
+    /**
+     * @var callable|string
+     */
+    private $canExpandCheck = null;
+
+    /**
+     * @var int
+     */
+    private $maxNestingLevel = null;
```
```suggestion
     */
    private static bool $default_max_nesting_level = 10;
    
    private string $name;

    private bool $expandNested = false;

    private bool $forceCloseNested = false;

    private GridField $gridField = null;

    private string $relationName = 'Children';

    private bool $inlineEditable = false;

    private callable|string $canExpandCheck = null;

    private int $maxNestingLevel = null;
```

ChloeSartorelli replied:
As per https://github.com/symbiote/silverstripe-gridfieldextensions/pull/384#discussion_r1573963964 these should be strict typehints - i.e. PHP should thrown errors at us if we set values that aren't those types.

forsdahl replied:
I changed these as requested, except the `$canExpandCallback` property. Seems like php doesn't like `callable` as a strict typehint on class properties, so I left that as a docblock typehint.

ChloeSartorelli commented on src/GridFieldNestedForm.php:
```diff
@@ -0,0 +1,517 @@
+    /**
+     * Set whether the nested grid fields should be inline editable
+     */
+    public function setInlineEditable(bool $editable)
```
```suggestion
    public function setInlineEditable(bool $editable): static
```
Please do this for all other `set*` methods which return `$this`

ChloeSartorelli commented on src/GridFieldNestedForm.php:
```diff
@@ -0,0 +1,517 @@
+     */
+    public function setCanExpandCheck(callable|string $callback)
+    {
+        $this->canExpandCheck = $callback;
```
```suggestion
        $this->canExpandCallback = $callback;
```
As per https://github.com/symbiote/silverstripe-gridfieldextensions/pull/384#discussion_r1573966303

ChloeSartorelli commented on src/GridFieldNestedForm.php:
```diff
@@ -0,0 +1,517 @@
+
+    /**
+     * Handle moving a record to a new parent
+     *
+     * @return string
+     */
+    public function handleMoveToParent(GridField $gridField, $request)
```
```suggestion
     */
    public function handleMoveToParent(GridField $gridField, $request): string
```
For all methods which are not inherited from a parent class or interface, please use strict typehints instead of phpdocs, e.g. handleNestedItem. I won't list them all.

ChloeSartorelli commented on src/GridFieldNestedForm.php:
```diff
@@ -0,0 +1,517 @@
+        ];
+    }
+
+    /**
+     * @param GridField $gridField
+     * @return array
+     */
```
This is defined on the interface, so it isn't needed here.
```suggestion
```

ChloeSartorelli commented on src/GridFieldNestedForm.php:
```diff
@@ -0,0 +1,517 @@
+     */
+    public function getHTMLFragments($gridField)
+    {
+        /**
```
```suggestion
        /*
```
The double asterisk tells IDEs that this comment may include typehints etc - but this is just a multi-line comment so it shouldn't have that.
I'm nitpicking here - I won't hold back merging if this change isn't made.

ChloeSartorelli commented on src/GridFieldNestedForm.php:
```diff
@@ -0,0 +1,517 @@
+        $list = $gridField->getList();
+        $id = isset($move['id']) ? (int) $move['id'] : null;
+        if (!$id) {
+            throw new HTTPResponse_Exception('Missing ID', 400);
```
```suggestion
            throw new HTTPResponse_Exception('Missing ID', 404);
```
404 if the item is not found

GuySartorelli reviewed (changes requested):
Thank you for your patience and diligent effort on this so far, we're very close now.

Please avoid marking comments as "resolved" - I use them to know what changes I requested last time (and have verified changes were made) vs which changes have already been dealt with in previous reviews.

There's also a minor PHP linting issue to resolve.

GuySartorelli commented:
In addition to the above, the following from https://github.com/symbiote/silverstripe-gridfieldextensions/pull/384#pullrequestreview-2013562592 haven't been resolved:

- There's some unexpected state conflict between the nested gridfield and the main gridfield (steps to reproduce in the original comment)
- Please add additional documentation so I know what needs to be manually tested and so that developers know what functionality this thing has.
- Please update the component list in the readme

GuySartorelli commented:
I've discovered that if you try to add two `GridFieldNestedForm` components, it doesn't work as expected (e.g. uncomment the second component in the PR description).
Probably we should just throw an exception if trying to add a `GridFieldNestedForm` when there's already one there. We can look at having multiple nested sibling gridfields working nicely together as a follow-up, if there's appetite for that.

forsdahl committed:
Linting and typehinting fixes for nested grid field

forsdahl committed:
Throw an exception in Nesed Gridfield if an invalid relation is configured.

forsdahl committed:
Only add nested form to nested gridfield child if that child is of
the same class as the parent record.

forsdahl committed:
Throw 404 error on grid field nested form move-to-parent action,
if no record is found.

forsdahl commented:
I looked in to the state handling issues you described, and it does seem to me to be the default state handling logic which isn't really up to handling this. The `GridFieldNestedForm` component doesn't do any custom logic for the state handling, neither in php nor in javascript, it relies on the gridfield state manager to add the correct state parameters for all links.
I can reproduce the scenario which you describe, and I can see that when having both a base grid field and a nested grid field with the `GridFieldFilterHeader` component, the state-part of the url is not updated properly when filtering.
I think the solution here is to update the general gridfield state handling to support this, with the `GridFieldStateManager` implementation we have which stores individual gridfield states in session, and only adds a state key to the url.
In the mean time, I think we could at least disable the `GridFieldFilterHeader` in the default nested gridfield config. In all the cases we have used nested gridfields, we have yet to see a real-world use case where we really feel the filtering in the nested gridfield to be really necessary and I suspect the state handling works ok if only the base gridfield has the `GridFieldFilterHeader`-component.

GuySartorelli commented:
There are still outstanding items I've requested above so I'll assume you're still working on this.
I've let CI run so you can see if that's green in the meantime.

GuySartorelli commented:
To find it easier to find the requested changes that haven't been made yet:
- https://github.com/symbiote/silverstripe-gridfieldextensions/pull/384#issuecomment-2089247085 regarding documentation and the readme
- https://github.com/symbiote/silverstripe-gridfieldextensions/pull/384#issuecomment-2089247995

forsdahl committed:
Added more documentation for nested gridfields.

forsdahl committed:
Only allow one GridFieldNestedForm component per GridField.

forsdahl commented:
Documentation and readme now updated.
GridFieldNestedForm now also throws an exception if it detects more than one instance of itself on the same GridField.

GuySartorelli reviewed (approved):
LGTM, nothing stands out as being obvious wrong in the code and it works great locally.
Thanks for your efforts on this, it's a really fantastic contribution!

This feature will be included in the October minor release.
